src/v2i-hub/ERVCloudForwardingPlugin/src/ERVCloudForwardingPlugin.cpp

```diff
@@ -0,0 +1,214 @@
+        // Subscribe to all messages specified by the filters above.
+        SubscribeToMessages();
+
+        std::thread webBSM_t(&ERVCloudForwardingPlugin::StartBSMWebService, this);
+        webBSM_t.detach();
+
+        // Send RSU location to cloud
+        RegisterRSULocation();
+    }
+
+    void ERVCloudForwardingPlugin::handleBSM(BsmMessage &msg, routeable_message &routableMsg)
+    {
+        // Check if the BSM is broadcast by an ERV (Emergency Response Vehicle)
+        if (ERVCloudForwardingWorker::IsBSMFromERV(msg))
+        {
+            // Construct the ERV BSM and forward it to the cloud.
+            auto xml_str = ERVCloudForwardingWorker::constructERVBSMRequest(msg);
+            PLOG(logINFO) << "Forward ERV BSM to cloud: " << xml_str << endl;
+            CloudSendAsync(xml_str, _CLOUDURL, _CLOUDBSMREQ, _POSTMETHOD);
+        }
+        else
+        {
+            // If BSM is not from ERV, print debug log
+            PLOG(logDEBUG) << "Incoming BSM is not from Emergency Response Vehicle (ERV)." << endl;
+        }
+    }
+
+    void ERVCloudForwardingPlugin::RegisterRSULocation()
+    {
+        try
+        {
+            PLOG(logINFO) << "Create SNMP Client to connect to RSU. RSU IP:" << _rsuIp << ",\tRSU Port:" << _snmpPort << ",\tSecurity Name:" << _securityUser << ",\tAuthentication Passphrase: " << _authPassPhrase << endl;
+            auto snmpClient = std::make_shared<SNMPClient>(_rsuIp, _snmpPort, _securityUser, _authPassPhrase);
+            auto gps_sentense = snmpClient->SNMPGet(_GPSOID);
+            auto gps_map = ERVCloudForwardingWorker::ParseGPS(gps_sentense);
+            long latitude = 0;
+            long longitude = 0;
+            for (auto itr = gps_map.begin(); itr != gps_map.end(); itr++)
+            {
+                latitude = itr->first;
+                longitude = itr->second;
+            }
+
+            if (latitude == 0 || longitude == 0)
+            {
+                PLOG(logERROR) << "Invalid latitude and longitude. Cannot register RSU location." << endl;
+                return;
+            }
+            auto uuid = boost::uuids::random_generator()();
+            string rsu_identifier = _rsuName + "_" + boost::lexical_cast<std::string>(uuid);
+            auto xml_str = ERVCloudForwardingWorker::constructRSULocationRequest(rsu_identifier, _webPort, latitude, longitude);
+            PLOG(logINFO) << "Sending registering RSU location reqest to cloud: " << xml_str << endl;
+            auto status = CloudSend(xml_str, _CLOUDURL, _CLOUDRSUREQ, _POSTMETHOD);
+            if (status == 1)
+            {
+                PLOG(logERROR) << "Cannot register RSU location. Reason: Failed to send RSU location to cloud." << endl;
+                return;
+            }
+        }
+        catch (SNMPClientException &ex)
+        {
+            PLOG(logERROR) << "Cannot register RSU location. Reason: " << ex.what() << endl;
+            return;
```

Comment:
Same as above

Reply:
The same idea as the above

Reply:
@dan-du-car Sorry, I'm late on this. Everything else looks good, but this is throwing errors from SNMPCleint() in SNMPClient.cpp, but it eventually is able to connect. Doesn't affect functionality, but do we want to look at fixing it?

Reply:
It is valid error if it does not connect.